github-actions/create-pr-for-changes/lib/main.ts

```diff
@@ -0,0 +1,204 @@
+    // files.
+    //
+    // NOTE:
+    //   Hashing the contents of the changed files is a heuristic to uniquely-ish represent a
+    //   particular set of changes. It is not 100% accurate (and could result in both false
+    //   positives and false negatives), but should be good enough for our purposes.
+    const branchPrefix = core.getInput('branch-prefix', {required: true});
+    const branchName = `${branchPrefix}-${hashFiles(touchedFiles)}`;
+
+    // Check whether there is a PR for the same changes already.
+    const {owner, repo} = context.repo;
+    const base = context.ref.replace(/^refs\/heads\//, '');
+
+    const {data: matchingPrs} = await git.github.pulls.list({
+      owner,
+      repo,
+      base,
+      head: `${owner}:${branchName}`,
+      state: 'open',
+    });
+
+    if (matchingPrs.length > 0) {
+      // A PR for the same set of changes does already exist. Exit.
+      core.info(
+        `Skipping PR creation, because there is already a PR: #${matchingPrs[0].number} ` +
+          `(${matchingPrs[0].html_url})`,
+      );
+      return;
+    } else {
+      core.info(`No pre-existing PR found for branch '${branchName}'.`);
+    }
+
+    // Gather PRs (if any) that will be superseded by the PR that will be created.
+    const {
+      data: {items: supersededPrs},
+    } = await git.github.search.issuesAndPullRequests({
+      q: `is:pull-request is:open repo:${owner}/${repo} base:${base} head:${branchPrefix}-`,
+      // NOTE:
+      //   We assume that there will be no more than 100 open PRs for the same type of change. This
+      //   is a reasonable assumption for our purposes, but even if it doesn't hold it won't be a
+      //   big problem: It just means that older PRs will not be auto-closed and will not be marked
+      //   as superseded.
+      per_page: 100,
+      sort: 'created',
+      order: 'desc',
+    });
+
+    core.info(
+      `Found ${supersededPrs.length} pre-existing PR(s) that will be superseded: ` +
+        (supersededPrs.length === 0 ? '-' : supersededPrs.map((pr) => `#${pr.number}`).join(', ')),
+    );
+
+    // Commit the current changes on a new branch.
+    const prTitle = core.getInput('pr-title', {required: true});
+    const prDescription = core.getInput('pr-description', {required: true});
+    const prFooter =
+      supersededPrs.length === 0
+        ? ''
+        : `\n${supersededPrs.map((pr) => `\nCloses #${pr.number}`).join('')}`;
+    const prBody = `${prDescription}${prFooter}`;
+    const prLabels = core
+      .getInput('pr-labels', {required: false})
+      .split('\n')
+      .map((l) => l.trim())
+      .filter((l) => l !== '');
+    const commitMessage = `${prTitle}\n\n${prBody}`;
+
+    git.run(['checkout', '-b', branchName]);
+    git.run(['add', '--all']);
+    git.run(['commit', '-m', commitMessage, '--no-verify']);
+    git.run(['push', '--force', git.getRepoGitUrl(), `HEAD:refs/heads/${branchName}`]);
```

Comment:
This shouldn't require a force if its a new branch.

Reply:
I used `--force`, because it is possible that the branch exists on the remote (for example, from a previous invocation of the action).

But I think it would be better to clean-up matching branches that do not correspond to open PRs (which will also make `--force` redundant).